examples/reshare/src/dkg/actor.rs

```diff
@@ -0,0 +1,138 @@
+            );
+
+            while let Some(message) = self.mailbox.next().await {
+                match message {
+                    Message::Act { response } => {
+                        let outcome = manager.take_deal_outcome();
+
+                        if let Some(ref outcome) = outcome {
+                            info!(
+                                n_acks = outcome.acks.len(),
+                                n_reveals = outcome.reveals.len(),
+                                "including reshare outcome in proposed block"
+                            );
+                        }
+
+                        let _ = response.send(outcome);
+                    }
+                    Message::Finalized { block } => {
+                        let round = block.height / EPOCH_LENGTH;
+                        let relative_height = block.height % EPOCH_LENGTH;
+
+                        // Finalize the round if at the end of an epoch.
+                        //
+                        // This will never include genesis, as marshal never reports that block.
+                        if relative_height == 0 {
+                            let Output { public, share } =
+                                manager.finalize(round.saturating_sub(1)).await;
+
+                            // Rotate the manager to begin a new round.
+                            manager = DkgManager::new(
+                                &mut self.context,
+                                public,
+                                share,
+                                &mut self.signer,
+                                &self.contributors,
+                                &mut sender,
+                                &mut receiver,
+                            );
+                        }
+
+                        match relative_height.cmp(&(EPOCH_LENGTH / 2)) {
+                            Ordering::Less => {
+                                // Continuously distribute shares to any players who haven't acknowledged
+                                // receipt yet.
+                                manager.distribute(round).await;
+
+                                // Process any incoming messages from other dealers/players.
+                                manager.process_messages(round).await;
+                            }
+                            Ordering::Equal => {
+                                // Process any final messages from other dealers/players.
+                                manager.process_messages(round).await;
+
+                                // At the midpoint of the epoch, construct the deal outcome for inclusion.
+                                manager.construct_deal_outcome(round);
+                            }
+                            Ordering::Greater => {
+                                // Process any incoming deal outcomes from dealing contributors.
+                                manager.process_block(round, block).await;
```

Comment:
I guess we could process blocks at all heights? But you aren't wrong to suggest we should consider those blocks invalid.